LuckParser/Controllers/StatisticsCalculator.cs

```diff
@@ -766,51 +766,70 @@ private void CalculateConditions()
         private void SetPresentBoons()
         {
             List<CombatItem> combatList = _log.GetCombatData();
+            var skillIDs = new HashSet<long>(combatList.Select(x => x.SkillID));
             if (_settings.PlayerBoonsUniversal)
-            {//Main boons
+            {
+                // Main boons
                 foreach (Boon boon in Boon.GetBoonList())
                 {
-                    if (combatList.Exists(x => x.SkillID == boon.GetID()))
+                    if (skillIDs.Contains(boon.GetID()))
                     {
                         _statistics.PresentBoons.Add(boon);
                     }
                 }
             }
+
             if (_settings.PlayerBoonsImpProf)
-            {//Important Class specefic boons
+            {
+                // Important class specific boons
                 foreach (Boon boon in Boon.GetOffensiveTableList())
                 {
-                    if (combatList.Exists(x => x.SkillID == boon.GetID()))
+                    if (skillIDs.Contains(boon.GetID()))
                     {
                         _statistics.PresentOffbuffs.Add(boon);
                     }
                 }
+
                 foreach (Boon boon in Boon.GetDefensiveTableList())
                 {
-                    if (combatList.Exists(x => x.SkillID == boon.GetID()))
+                    if (skillIDs.Contains(boon.GetID()))
                     {
                         _statistics.PresentDefbuffs.Add(boon);
                     }
                 }
             }
 
-            foreach (Player p in _log.GetPlayerList())
+            if (_settings.PlayerBoonsAllProf)
             {
-                _statistics.PresentPersonnalBuffs[p.GetInstid()] = new List<Boon>();
-                if (_settings.PlayerBoonsAllProf)
-                {//All class specefic boons
-                    List<Boon> notYetFoundBoons = Boon.GetRemainingBuffsList();
-                    combatList.ForEach(item =>
+                var players = _log.GetPlayerList();
+                var playersById = new Dictionary<ushort, Player>();
+                foreach (var player in players)
+                {
+                    _statistics.PresentPersonnalBuffs[player.GetInstid()] = new List<Boon>();
```

Comment:
If we are creating this list under condition, some additional checks must be made in statisticsCalculator for each "boonToTrack.AddRange(_statistics.PresentPersonnalBuffs[player.GetInstid()]);" htmlbuilder seems to do a ContainsKey before trying to use it but that's not the case in statistics calculator

Reply:
Seems to me like the proper fix here is to add that ContainsKey check or make sure you always have a list there for each player (which would be *significantly* more complicated as that would involve rewriting pretty much the whole Statistics class to do it properly, but it would make it way easier to avoid this kind of things).

What do you think?

Reply:
For the scope of this PR I think ContainsKey s should be enough. I'm seeing that it's done four times but it can only be done once (there is 4 consecutive foreach(Players in playerList) that could easily be put in a single foreach)